src/client/coll/mod.rs

```diff
@@ -254,95 +254,165 @@ impl<'a> Collection<'a> {
-        }).collect();
+    fn execute_update_batch(&self, models: Vec<UpdateModel>, start_index: i64,
+                            ordered: bool, result: &mut BulkWriteResult,
+                            exception: &mut BulkWriteException) -> bool{
+        let original_models = models.iter().map(|model|
+            if model.multi {
+                WriteModel::DeleteMany { filter: model.filter.clone() }
+            } else {
+                WriteModel::DeleteOne { filter: model.filter.clone() }
+            }
+        ).collect();
 
-        match self.insert_many(documents, ordered, None) {
-            Ok(insert_result) =>
-                result.process_insert_many_result(insert_result, models,
+        match self.bulk_update(models, ordered, None) {
+            Ok(bulk_update_result) =>
+                result.process_bulk_update_result(bulk_update_result,
+                                                  original_models, start_index,
                                                   exception),
-            Err(err) => exception.add_unproccessed_models(models)
+            Err(_) =>  {
+                exception.add_unproccessed_models(original_models);
+                false
+            }
         }
     }
 
-    fn execute_batch(&self, batch: Batch, ordered: bool, i: i64,
+    fn execute_batch(&self, batch: Batch, start_index: i64, ordered: bool,
                      result: &mut BulkWriteResult,
-                     exception: &mut BulkWriteException) {
+                     exception: &mut BulkWriteException) -> bool {
         match batch {
-            Batch::Insert { mut documents } =>
-                if documents.len() == 1 {
-                    self.execute_insert_one_batch(documents.pop().unwrap(), i, result,
-                                                  exception)
-                } else {
-                    self.execute_insert_many_batch(documents, ordered,
-                                                   result, exception)
-                }
+            Batch::Insert(docs) =>
+                self.execute_insert_batch(docs, start_index, ordered, result,
+                                          exception),
+            Batch::Delete(models) =>
+                self.execute_delete_batch(models, ordered, result,
+                                          exception),
+            Batch::Update(models) =>
+                self.execute_update_batch(models, start_index, ordered, result,
+                                          exception),
         }
     }
 
     /// Sends a batch of writes to the server at the same time.
     pub fn bulk_write(&self, requests: Vec<WriteModel>, ordered: bool) -> BulkWriteResult {
         let batches = if ordered {
-            Collection::get_ordered_batches(requests)
-        } else {
-            Collection::get_unordered_batches(requests)
-        };
+                          Collection::get_ordered_batches(VecDeque::from_iter(requests.into_iter()))
```

Comment:
I checked the official Rust repo, and it seems that the unindented style is what they use, so I'll stick with that (see line 530 of https://github.com/rust-lang/rust/blob/f76d9bcfc2c269452522fbbe19f66fe653325646/src/libsyntax/ext/build.rs)